**Attempt 2**


---

Optimization Techniques:

*   Dropped fewer columns - This time NAME will not be dropped. - From Attempt 1
*   Created bins for rare occurrences in NAME column. - From Attempt 1
*  Created bins for rare occurrences in ASK_AMT column.



## Preprocessing

In [ ]:
# Import our dependencies
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
import pandas as pd
import tensorflow as tf

#  Import and read the charity_data.csv.
import pandas as pd
application_df = pd.read_csv("https://static.bc-edx.com/data/dl-1-2/m21/lms/starter/charity_data.csv")
application_df.head()

EIN                                      NAME APPLICATION_TYPE  \
0  10520599              BLUE KNIGHTS MOTORCYCLE CLUB              T10   
1  10531628    AMERICAN CHESAPEAKE CLUB CHARITABLE TR               T3   
2  10547893        ST CLOUD PROFESSIONAL FIREFIGHTERS               T5   
3  10553066            SOUTHSIDE ATHLETIC ASSOCIATION               T3   
4  10556103  GENETIC RESEARCH INSTITUTE OF THE DESERT               T3   

        AFFILIATION CLASSIFICATION      USE_CASE  ORGANIZATION  STATUS  \
0       Independent          C1000    ProductDev   Association       1   
1       Independent          C2000  Preservation  Co-operative       1   
2  CompanySponsored          C3000    ProductDev   Association       1   
3  CompanySponsored          C2000  Preservation         Trust       1   
4       Independent          C1000     Heathcare         Trust       1   

      INCOME_AMT SPECIAL_CONSIDERATIONS  ASK_AMT  IS_SUCCESSFUL  
0              0                      N     5000              1  
1         1-9999                      N   108590              1  
2              0                      N     5000              0  
3    10000-24999                      N     6692              1  
4  100000-499999                      N   142590              1

In [ ]:
# Drop the non-beneficial ID column 'EIN'.
application_df = application_df.drop(columns=['EIN'])

In [ ]:
# Determine the number of unique values in each column.
print(application_df.nunique())

NAME                      19568
APPLICATION_TYPE             17
AFFILIATION                   6
CLASSIFICATION               71
USE_CASE                      5
ORGANIZATION                  4
STATUS                        2
INCOME_AMT                    9
SPECIAL_CONSIDERATIONS        2
ASK_AMT                    8747
IS_SUCCESSFUL                 2
dtype: int64


In [ ]:
# Look at APPLICATION_TYPE value counts to identify and replace with "Other"
print(application_df['APPLICATION_TYPE'].value_counts())

APPLICATION_TYPE
T3     27037
T4      1542
T6      1216
T5      1173
T19     1065
T8       737
T7       725
T10      528
T9       156
T13       66
T12       27
T2        16
T25        3
T14        3
T29        2
T15        2
T17        1
Name: count, dtype: int64


In [ ]:
# Choose a cutoff value and create a list of application types to be replaced
# use the variable name `application_types_to_replace`
application_types_to_replace = application_df['APPLICATION_TYPE'].value_counts()[application_df['APPLICATION_TYPE'].value_counts() < 500].index.tolist()

# Replace in dataframe
for app in application_types_to_replace:
    application_df['APPLICATION_TYPE'] = application_df['APPLICATION_TYPE'].replace(app,"Other")

# Check to make sure replacement was successful
application_df['APPLICATION_TYPE'].value_counts()

APPLICATION_TYPE
T3       27037
T4        1542
T6        1216
T5        1173
T19       1065
T8         737
T7         725
T10        528
Other      276
Name: count, dtype: int64

In [ ]:
# Look at CLASSIFICATION value counts to identify and replace with "Other"
print(application_df['CLASSIFICATION'].value_counts())

CLASSIFICATION
C1000    17326
C2000     6074
C1200     4837
C3000     1918
C2100     1883
         ...  
C4120        1
C8210        1
C2561        1
C4500        1
C2150        1
Name: count, Length: 71, dtype: int64


In [ ]:
# You may find it helpful to look at CLASSIFICATION value counts >1
classification_value_counts = application_df['CLASSIFICATION'].value_counts()
filtered_classification_value_counts = classification_value_counts[classification_value_counts > 1]

print(filtered_classification_value_counts)

CLASSIFICATION
C1000    17326
C2000     6074
C1200     4837
C3000     1918
C2100     1883
C7000      777
C1700      287
C4000      194
C5000      116
C1270      114
C2700      104
C2800       95
C7100       75
C1300       58
C1280       50
C1230       36
C1400       34
C7200       32
C2300       32
C1240       30
C8000       20
C7120       18
C1500       16
C1800       15
C6000       15
C1250       14
C8200       11
C1238       10
C1278       10
C1235        9
C1237        9
C7210        7
C2400        6
C1720        6
C4100        6
C1257        5
C1600        5
C1260        3
C2710        3
C0           3
C3200        2
C1234        2
C1246        2
C1267        2
C1256        2
Name: count, dtype: int64


In [ ]:
# Choose a cutoff value and create a list of classifications to be replaced
# use the variable name `classifications_to_replace`
classifications_to_replace = application_df['CLASSIFICATION'].value_counts()[application_df['CLASSIFICATION'].value_counts() < 1000].index.tolist()

# Replace in dataframe
for cls in classifications_to_replace:
    application_df['CLASSIFICATION'] = application_df['CLASSIFICATION'].replace(cls,"Other")

# Check to make sure replacement was successful
application_df['CLASSIFICATION'].value_counts()

CLASSIFICATION
C1000    17326
C2000     6074
C1200     4837
Other     2261
C3000     1918
C2100     1883
Name: count, dtype: int64

In [ ]:
# Look at NAME value counts to identify and replace with "Other"
print(application_df['NAME'].value_counts())

NAME
PARENT BOOSTER USA INC                                                  1260
TOPS CLUB INC                                                            765
UNITED STATES BOWLING CONGRESS INC                                       700
WASHINGTON STATE UNIVERSITY                                              492
AMATEUR ATHLETIC UNION OF THE UNITED STATES INC                          408
                                                                        ... 
ST LOUIS SLAM WOMENS FOOTBALL                                              1
AIESEC ALUMNI IBEROAMERICA CORP                                            1
WEALLBLEEDRED ORG INC                                                      1
AMERICAN SOCIETY FOR STANDARDS IN MEDIUMSHIP & PSYCHICAL INVESTIGATI       1
WATERHOUSE CHARITABLE TR                                                   1
Name: count, Length: 19568, dtype: int64


In [ ]:
# You may find it helpful to look at NAME value counts >1
name_value_counts = application_df['NAME'].value_counts()
filtered_name_value_counts = name_value_counts[name_value_counts > 1]

print(filtered_name_value_counts)

NAME
PARENT BOOSTER USA INC                                            1260
TOPS CLUB INC                                                      765
UNITED STATES BOWLING CONGRESS INC                                 700
WASHINGTON STATE UNIVERSITY                                        492
AMATEUR ATHLETIC UNION OF THE UNITED STATES INC                    408
                                                                  ... 
VETERANS OF FOREIGH WARS OF THE UNITED STATES DEPT OF COLORADO       2
ETA PHI BETA SORORITY INC                                            2
POINT MAN INTERNATIONAL MINISTRIES                                   2
MULTI COMMUNITY DIVERSIFIED SERVICES INC                             2
VETERANS OF FOREIGN WARS AUXILIARY                                   2
Name: count, Length: 792, dtype: int64


In [ ]:
# You may find it helpful to look at NAME value counts >2
name_value_counts_2 = application_df['NAME'].value_counts()
filtered_name_value_counts_2 = name_value_counts_2[name_value_counts_2 > 2]

print(filtered_name_value_counts_2)

NAME
PARENT BOOSTER USA INC                                             1260
TOPS CLUB INC                                                       765
UNITED STATES BOWLING CONGRESS INC                                  700
WASHINGTON STATE UNIVERSITY                                         492
AMATEUR ATHLETIC UNION OF THE UNITED STATES INC                     408
                                                                   ... 
ENTREPRENEURS ORGANIZATION                                            3
CENTER FOR THE STUDY OF CLASSICAL ARCHITECTURE INC                    3
VETERANS OF FOREIGN WARS OF THE UNITED STATES DEPT OF TENNESSEE       3
AMERICAN MARKETING ASSOCIATION INC                                    3
UNITED STATES BOWLING CONGRESS                                        3
Name: count, Length: 577, dtype: int64


In [ ]:
# You may find it helpful to look at NAME value counts >5
name_value_counts_5 = application_df['NAME'].value_counts()
filtered_name_value_counts_5 = name_value_counts_5[name_value_counts_5 > 5]

print(filtered_name_value_counts_5)

NAME
PARENT BOOSTER USA INC                             1260
TOPS CLUB INC                                       765
UNITED STATES BOWLING CONGRESS INC                  700
WASHINGTON STATE UNIVERSITY                         492
AMATEUR ATHLETIC UNION OF THE UNITED STATES INC     408
                                                   ... 
OLD OAK CLIFF CONSERVATION LEAGUE INC                 6
AMERICAN NEPHROLOGY NURSES ASSOCIATION                6
HUMBLE ISD EDUCATIONAL SUPPORT GROUPS INC             6
PROFESSIONAL LOADMASTER ASSOCIATION                   6
CBMC INC                                              6
Name: count, Length: 354, dtype: int64


In [ ]:
# You may find it helpful to look at NAME value counts >100
name_value_counts_100 = application_df['NAME'].value_counts()
filtered_name_value_counts_100 = name_value_counts_100[name_value_counts_100 > 100]

print(filtered_name_value_counts_100)

NAME
PARENT BOOSTER USA INC                                                1260
TOPS CLUB INC                                                          765
UNITED STATES BOWLING CONGRESS INC                                     700
WASHINGTON STATE UNIVERSITY                                            492
AMATEUR ATHLETIC UNION OF THE UNITED STATES INC                        408
PTA TEXAS CONGRESS                                                     368
SOROPTIMIST INTERNATIONAL OF THE AMERICAS INC                          331
ALPHA PHI SIGMA                                                        313
TOASTMASTERS INTERNATIONAL                                             293
MOST WORSHIPFUL STRINGER FREE AND ACCEPTED MASONS                      287
LITTLE LEAGUE BASEBALL INC                                             277
INTERNATIONAL ASSOCIATION OF LIONS CLUBS                               266
MOMS CLUB                                                              210
INTERNATIONAL ASSOCI

In [ ]:
# Choose a cutoff value and create a list of names to be replaced
# use the variable name `names_to_replace`
names_to_replace = application_df['NAME'].value_counts()[application_df['NAME'].value_counts() < 300].index.tolist()

# Replace in dataframe
for cls in names_to_replace:
    application_df['NAME'] = application_df['NAME'].replace(cls,"Other")

# Check to make sure replacement was successful
application_df['NAME'].value_counts()

NAME
Other                                              29662
PARENT BOOSTER USA INC                              1260
TOPS CLUB INC                                        765
UNITED STATES BOWLING CONGRESS INC                   700
WASHINGTON STATE UNIVERSITY                          492
AMATEUR ATHLETIC UNION OF THE UNITED STATES INC      408
PTA TEXAS CONGRESS                                   368
SOROPTIMIST INTERNATIONAL OF THE AMERICAS INC        331
ALPHA PHI SIGMA                                      313
Name: count, dtype: int64

In [ ]:
# Look at ASK_AMT value counts
print(application_df['ASK_AMT'].value_counts())

ASK_AMT
5000        25398
10478           3
15583           3
63981           3
6725            3
            ...  
5371754         1
30060           1
43091152        1
18683           1
36500179        1
Name: count, Length: 8747, dtype: int64


In [ ]:
application_df["ASK_AMT"].describe()

count    3.429900e+04
mean     2.769199e+06
std      8.713045e+07
min      5.000000e+03
25%      5.000000e+03
50%      5.000000e+03
75%      7.742000e+03
max      8.597806e+09
Name: ASK_AMT, dtype: float64

In [ ]:
# Define custom bins and labels for ASK_AMT
bins = [0, 5000, 7742, float('inf')]  # Define the thresholds for Low, Medium, High
labels = ['Low', 'Medium', 'High']       # Labels for the bins

# Apply custom binning to ASK_AMT
application_df['ASK_AMT_Binned'] = pd.cut(application_df['ASK_AMT'], bins=bins, labels=labels, include_lowest=True)

# Display the resulting DataFrame to check the new binned column
application_df.head(20)


NAME APPLICATION_TYPE       AFFILIATION CLASSIFICATION      USE_CASE  \
0   Other              T10       Independent          C1000    ProductDev   
1   Other               T3       Independent          C2000  Preservation   
2   Other               T5  CompanySponsored          C3000    ProductDev   
3   Other               T3  CompanySponsored          C2000  Preservation   
4   Other               T3       Independent          C1000     Heathcare   
5   Other               T3       Independent          C1200  Preservation   
6   Other               T3       Independent          C1000  Preservation   
7   Other               T3       Independent          C2000  Preservation   
8   Other               T7       Independent          C1000    ProductDev   
9   Other               T5  CompanySponsored          C3000    ProductDev   
10  Other               T3       Independent          C1200  Preservation   
11  Other               T3       Independent          C2000  Preservation   
12  Other               T3  CompanySponsored          C1200  Preservation   
13  Other               T3       Independent          Other  Preservation   
14  Other               T3       Independent          C1000  Preservation   
15  Other               T3       Independent          C1200  Preservation   
16  Other               T3       Independent          C2000  Preservation   
17  Other               T3       Independent          C1000     Heathcare   
18  Other               T3  CompanySponsored          C1000  Preservation   
19  Other               T3  CompanySponsored          C1000  Preservation   

    ORGANIZATION  STATUS     INCOME_AMT SPECIAL_CONSIDERATIONS   ASK_AMT  \
0    Association       1              0                      N      5000   
1   Co-operative       1         1-9999                      N    108590   
2    Association       1              0                      N      5000   
3          Trust       1    10000-24999                      N      6692   
4          Trust       1  100000-499999                      N    142590   
5          Trust       1              0                      N      5000   
6          Trust       1  100000-499999                      N     31452   
7          Trust       1        10M-50M                      N   7508025   
8          Trust       1         1-9999                      N     94389   
9    Association       1              0                      N      5000   
10         Trust       1    25000-99999                      N     69656   
11         Trust       1  100000-499999                      N    165593   
12   Association       1              0                      N      5000   
13         Trust       1    25000-99999                      N      5301   
14         Trust       1              0                      N      5000   
15         Trust       1              0                      N      5000   
16         Trust       1              0                      N      5000   
17         Trust       1           50M+                      N  86380556   
18         Trust       1              0                      N      5000   
19         Trust       1  100000-499999                      N    376109   

    IS_SUCCESSFUL ASK_AMT_Binned  
0               1            Low  
1               1           High  
2               0            Low  
3               1         Medium  
4               1           High  
5               1            Low  
6               1           High  
7               1           High  
8               1           High  
9               0            Low  
10              0           High  
11              0           High  
12              1            Low  
13              1         Medium  
14              1            Low  
15              1            Low  
16              1            Low  
17              0           High  
18              0            Low  
19              0           High

In [ ]:
# Drop the 'ASK_AMT'.
application_df = application_df.drop(columns=['ASK_AMT'])

# Display the updated DataFrame
application_df.head()

NAME APPLICATION_TYPE       AFFILIATION CLASSIFICATION      USE_CASE  \
0  Other              T10       Independent          C1000    ProductDev   
1  Other               T3       Independent          C2000  Preservation   
2  Other               T5  CompanySponsored          C3000    ProductDev   
3  Other               T3  CompanySponsored          C2000  Preservation   
4  Other               T3       Independent          C1000     Heathcare   

   ORGANIZATION  STATUS     INCOME_AMT SPECIAL_CONSIDERATIONS  IS_SUCCESSFUL  \
0   Association       1              0                      N              1   
1  Co-operative       1         1-9999                      N              1   
2   Association       1              0                      N              0   
3         Trust       1    10000-24999                      N              1   
4         Trust       1  100000-499999                      N              1   

  ASK_AMT_Binned  
0            Low  
1           High  
2            Low  
3         Medium  
4           High

In [ ]:
print(application_df.dtypes)

NAME                        object
APPLICATION_TYPE            object
AFFILIATION                 object
CLASSIFICATION              object
USE_CASE                    object
ORGANIZATION                object
STATUS                       int64
INCOME_AMT                  object
SPECIAL_CONSIDERATIONS      object
IS_SUCCESSFUL                int64
ASK_AMT_Binned            category
dtype: object


In [ ]:
# Identify the categorical columns in your DataFrame
categorical_columns = ['NAME', 'APPLICATION_TYPE', 'AFFILIATION', 'CLASSIFICATION', 'USE_CASE', 'ORGANIZATION', 'INCOME_AMT', 'SPECIAL_CONSIDERATIONS', 'ASK_AMT_Binned']

# Convert categorical data to numeric with `pd.get_dummies`
application_encoded = pd.get_dummies(application_df[categorical_columns])

In [ ]:
# Concatenate the application_df and the application_encoded DataFrames
application_df = pd.concat([application_df, application_encoded], axis=1)

application_df = application_df.drop(categorical_columns, axis=1)

# Display sample data
application_df.head()

STATUS  IS_SUCCESSFUL  NAME_ALPHA PHI SIGMA  \
0       1              1                 False   
1       1              1                 False   
2       1              0                 False   
3       1              1                 False   
4       1              1                 False   

   NAME_AMATEUR ATHLETIC UNION OF THE UNITED STATES INC  NAME_Other  \
0                                              False           True   
1                                              False           True   
2                                              False           True   
3                                              False           True   
4                                              False           True   

   NAME_PARENT BOOSTER USA INC  NAME_PTA TEXAS CONGRESS  \
0                        False                    False   
1                        False                    False   
2                        False                    False   
3                        False                    False   
4                        False                    False   

   NAME_SOROPTIMIST INTERNATIONAL OF THE AMERICAS INC  NAME_TOPS CLUB INC  \
0                                              False                False   
1                                              False                False   
2                                              False                False   
3                                              False                False   
4                                              False                False   

   NAME_UNITED STATES BOWLING CONGRESS INC  ...  INCOME_AMT_10M-50M  \
0                                    False  ...               False   
1                                    False  ...               False   
2                                    False  ...               False   
3                                    False  ...               False   
4                                    False  ...               False   

   INCOME_AMT_1M-5M  INCOME_AMT_25000-99999  INCOME_AMT_50M+  \
0             False                   False            False   
1             False                   False            False   
2             False                   False            False   
3             False                   False            False   
4             False                   False            False   

   INCOME_AMT_5M-10M  SPECIAL_CONSIDERATIONS_N  SPECIAL_CONSIDERATIONS_Y  \
0              False                      True                     False   
1              False                      True                     False   
2              False                      True                     False   
3              False                      True                     False   
4              False                      True                     False   

   ASK_AMT_Binned_Low  ASK_AMT_Binned_Medium  ASK_AMT_Binned_High  
0                True                  False                False  
1               False                  False                 True  
2                True                  False                False  
3               False                   True                False  
4               False                  False                 True  

[5 rows x 55 columns]

In [ ]:
# Split our preprocessed data into our features and target arrays
y = application_df["IS_SUCCESSFUL"].values
X = application_df.drop(["IS_SUCCESSFUL"], axis=1).values

# Split the preprocessed data into a training and testing dataset
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=78)

In [ ]:
# Create a StandardScaler instances
scaler = StandardScaler()

# Fit the StandardScaler
X_scaler = scaler.fit(X_train)

# Scale the data
X_train_scaled = X_scaler.transform(X_train)
X_test_scaled = X_scaler.transform(X_test)

## Compile, Train and Evaluate the Model

In [ ]:
# Define the model - deep neural net, i.e., the number of input features and hidden nodes for each layer.
input_features = len(X_train[0])
hidden_layer_units_1 =  80
hidden_layer_units_2 = 30

nn = tf.keras.models.Sequential()

# First hidden layer
nn.add(tf.keras.layers.Dense(units=hidden_layer_units_1, activation="relu", input_dim=input_features))

# Second hidden layer
nn.add(tf.keras.layers.Dense(units=hidden_layer_units_2, activation="relu"))

# Output layer
nn.add(tf.keras.layers.Dense(units=1, activation="sigmoid"))

# Check the structure of the model
nn.summary()

/usr/local/lib/python3.10/dist-packages/keras/src/layers/core/dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Model: "sequential"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━┓
┃ Layer (type)                         ┃ Output Shape                ┃         Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━┩
│ dense (Dense)                        │ (None, 80)                  │           4,400 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ dense_1 (Dense)                      │ (None, 30)                  │           2,430 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ dense_2 (Dense)                      │ (None, 1)                   │              31 │
└──────────────────────────────────────┴─────────────────────────────┴─────────────────┘

 Total params: 6,861 (26.80 KB)

 Trainable params: 6,861 (26.80 KB)

 Non-trainable params: 0 (0.00 B)

In [ ]:
# Compile the model
nn.compile(loss="binary_crossentropy", optimizer="adam", metrics=["accuracy"])

In [ ]:
# Train the model
fit_model = nn.fit(X_train_scaled, y_train, epochs=100)

Epoch 1/100
804/804 ━━━━━━━━━━━━━━━━━━━━ 3s 2ms/step - accuracy: 0.7140 - loss: 0.5570
Epoch 2/100
804/804 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.7420 - loss: 0.5242
Epoch 3/100
804/804 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.7436 - loss: 0.5207
Epoch 4/100
804/804 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.7484 - loss: 0.5147
Epoch 5/100
804/804 ━━━━━━━━━━━━━━━━━━━━ 3s 3ms/step - accuracy: 0.7429 - loss: 0.5200
Epoch 6/100
804/804 ━━━━━━━━━━━━━━━━━━━━ 4s 2ms/step - accuracy: 0.7411 - loss: 0.5187
Epoch 7/100
804/804 ━━━━━━━━━━━━━━━━━━━━ 3s 2ms/step - accuracy: 0.7466 - loss: 0.5143
Epoch 8/100
804/804 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.7476 - loss: 0.5131
Epoch 9/100
804/804 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.7466 - loss: 0.5130
Epoch 10/100
804/804 ━━━━━━━━━━━━━━━━━━━━ 3s 3ms/step - accuracy: 0.7474 - loss: 0.5136
Epoch 11/100
804/804 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.7407 - loss: 0.5127
Epoch 12/100
804/804 ━━━━━━━━━━━━━━━━━━━━

In [ ]:
# Evaluate the model using the test data
model_loss, model_accuracy = nn.evaluate(X_test_scaled,y_test,verbose=2)
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

268/268 - 0s - 2ms/step - accuracy: 0.7387 - loss: 0.5269
Loss: 0.5268723964691162, Accuracy: 0.7386589050292969


**The target predictive accuracy decreased sligltly from 0.7388 (Attempt 1) to 0.7387 (Attempt 2).**

In [ ]:
# # Export our model to HDF5 file
# nn.save("AlphabetSoupCharity_Optimization.h5")